src/backend/distributed/commands/database.c

```diff
@@ -242,3 +252,159 @@ PreprocessAlterDatabaseSetStmt(Node *node, const char *queryString,
+
+	StringInfo internalCreateCommand = makeStringInfo();
+	appendStringInfo(internalCreateCommand,
+					 "SELECT pg_catalog.citus_internal_database_command(%s)",
+					 quote_literal_cstr(createDatabaseCommand));
+
+	List *commands = list_make3(DISABLE_DDL_PROPAGATION,
+								(void *) internalCreateCommand->data,
+								ENABLE_DDL_PROPAGATION);
+
+	return NodeDDLTaskList(NON_COORDINATOR_NODES, commands);
+}
+
+
+/*
+ * citus_internal_database_command is an internal UDF to
+ * create/drop a database in an idempotent maner without
+ * transaction block restrictions.
+ */
+Datum
+citus_internal_database_command(PG_FUNCTION_ARGS)
+{
+	int saveNestLevel = NewGUCNestLevel();
+	text *commandText = PG_GETARG_TEXT_P(0);
+	char *command = text_to_cstring(commandText);
+	Node *parseTree = ParseTreeNode(command);
+
+	set_config_option("citus.enable_ddl_propagation", "off",
+					  (superuser() ? PGC_SUSET : PGC_USERSET), PGC_S_SESSION,
+					  GUC_ACTION_LOCAL, true, 0, false);
+
+	set_config_option("citus.enable_create_database_propagation", "off",
+					  (superuser() ? PGC_SUSET : PGC_USERSET), PGC_S_SESSION,
+					  GUC_ACTION_LOCAL, true, 0, false);
+
+	/*
+	 * createdb() / DropDatabase() uses ParseState to report the error position for the
+	 * input command and the position is reported to be 0 when it's provided as NULL.
+	 * We're okay with that because we don't expect this UDF to be called with an incorrect
+	 * DDL command.
+	 *
+	 */
+	ParseState *pstate = NULL;
+
+	if (IsA(parseTree, CreatedbStmt))
+	{
+		CreatedbStmt *stmt = castNode(CreatedbStmt, parseTree);
+
+		bool missingOk = true;
+		Oid databaseOid = get_database_oid(stmt->dbname, missingOk);
+
+		if (!OidIsValid(databaseOid))
+		{
+			createdb(pstate, (CreatedbStmt *) parseTree);
+		}
+	}
+	else if (IsA(parseTree, DropdbStmt))
+	{
+		DropdbStmt *stmt = castNode(DropdbStmt, parseTree);
+
+		bool missingOk = false;
+		Oid databaseOid = get_database_oid(stmt->dbname, missingOk);
+
+
+		if (OidIsValid(databaseOid))
+		{
+			DropDatabase(pstate, (DropdbStmt *) parseTree);
+		}
+	}
+	else
+	{
+		ereport(ERROR, (errmsg("unsupported command type %d", nodeTag(parseTree))));
+	}
+
+	/* Below command rollbacks flags to the state before this session*/
+	AtEOXact_GUC(true, saveNestLevel);
+
+	PG_RETURN_VOID();
+}
+
+
+List *
+PreprocessDropDatabaseStmt(Node *node, const char *queryString,
+						   ProcessUtilityContext processUtilityContext)
+{
+	if (!EnableCreateDatabasePropagation || !ShouldPropagate())
+	{
+		return NIL;
+	}
+
+	EnsureCoordinator();
+
+	DropdbStmt *stmt = (DropdbStmt *) node;
+
+	Oid databaseOid = get_database_oid(stmt->dbname, stmt->missing_ok);
+
+	if (databaseOid == InvalidOid)
+	{
+		/* let regular ProcessUtility deal with IF NOT EXISTS */
+		return NIL;
+	}
+
+	ObjectAddress dbAddress = { 0 };
+	ObjectAddressSet(dbAddress, DatabaseRelationId, databaseOid);
+	if (!IsObjectDistributed(&dbAddress))
+	{
+		return NIL;
+	}
+
+	UnmarkObjectDistributed(&dbAddress);
```

Comment:
This deletes the object only from local pg_dist_object, need to do so on other nodes as well.

What do we do to clean-up pg_dist_object in case of other "DROP foo_bar" type of commands?

Reply:
We don't seem to care about that for other distributed object types too, so nvm.

But I'm really wondering how that doesn't cause any sort of issues, cc: @thanodnl.

Reply:
@onurctirtir I'm checking other commands as well. I checked for drop role and it is being deleted from worker nodes. Checking for other commands as well

Reply:
>  I checked for drop role and it is being deleted from worker nodes.

Could you share the code that accomplishes this?

Reply:
create role role1;

select * from pg_dist_objects;

drop role role1;

select * from pg_dist_objects;